testnet/stacks-node/src/neon_node.rs

```diff
@@ -228,22 +229,73 @@ fn spawn_peer(mut this: PeerNetwork, p2p_sock: &SocketAddr, rpc_sock: &SocketAdd
         false, TESTNET_CHAIN_ID, &stacks_chainstate_path)
         .map_err(NetError::DBError)?;
 
+    // buffer up blocks to store without stalling the p2p thread
+    let mut results_with_data = VecDeque::new();
+    let mut wokeup_relayer = false;
+
     let server_thread = thread::spawn(move || {
         loop {
-            let network_result = this.run(&burndb, &mut chainstate, &mut mem_pool, Some(&mut dns_client), poll_timeout)
+            let download_backpressure = results_with_data.len() > 0;
+            let has_attacheable_blocks = StacksChainState::has_attacheable_staging_blocks(&chainstate.blocks_db).expect("BUG: failed to query available staging blocks");
```

Comment:
Why does the p2p thread need to poll the Stacks chain state for attachable blocks?

Reply:
The relayer thread needs to get woken up in order to call `process_blocks()`, which will process up to a given number of attachable blocks.  Right now, the relayer thread is driven entirely by its communication channel from the network thread, so as a short-term hack I made it so the network thread will use the channel to wake up the relayer when there's work for it to do.

I'm not satisfied with this, however.  I'll make it so the relayer thread just does a `try_recv()` to get data, and if there's no new data, it just calls `process_blocks(1)` on receive failure.

Reply:
That seems less than ideal as well -- the relayer thread really should only wake when it receives an event on it's channel. Shouldn't this just be an event fired from p2p.run (which is where the blocks are moving to the attachable state, right?)

Reply:
> Shouldn't this just be an event fired from p2p.run (which is where the blocks are moving to the attachable state, right?)

Actually no.  Blocks are marked as attachable once they're stored to staging, which happens within `process_blocks()`, which is in turn called by the relayer's `process_network_result()` method.  The relayer will always process up to `new_blocks.len() + C` blocks per call to `process_network_result()`, and those blocks will either be marked as orphaned or will be marked as included in the chainstate (either way, they're no longer attachable).

I'd like to keep the p2p thread as responsive as possible to network requests, which is why I thought it would be appropriate to have the relayer deciding when it's okay to process more blocks.

Reply:
So, the case this is trying to address is the case where the relayer thread has called `process_blocks`, but that call didn't finish processing all the possible processable blocks, so it needs to call `process_blocks` again?

Reply:
> So, the case this is trying to address is the case where the relayer thread has called process_blocks, but that call didn't finish processing all the possible processable blocks, so it needs to call process_blocks again?

Yes, but there's more to this that this snippit of code is addressing.  What happens now, without this patch, is the p2p network will download blocks faster than the relayer can process them.  What ultimately happens is the p2p network ends up dropping the downloaded blocks because the channel gets too full of network results (this happens even if we only pass along network results that contain blocks).  This, in turn, causes the p2p network to re-download blocks needlessly, because until the relayer gets around to storing them, the p2p network will believe that it still needs to get them.  Moreover, dropping blocks slows the node's block-processing down, since it won't be able to process any blocks after the dropped ones until the downloader gets around to discovering that the blocks were missing (i.e. the downloader does a best-effort sweep of all winning sortitions, and only comes back to re-trying failed downloads on the next sweep).

At the same time, we can't synchronously send the blocks to the relayer either, since this will just cause the p2p network thread to stall once the channel is full.  But, if we know that the receiver channel is not full (i.e. there's no backpressure), then the p2p network thread should be fetching _new_ blocks as quickly as possible.

What this code is trying to do is making it so the p2p code fetches blocks as quickly as possible while there is no backpressure, but cease fetching blocks once there is backpressure (but continue to process all other network requests).  In addition, if the p2p code receives a block, microblock, or transaction while there is backpressure, or if backpressure arises in the middle of doing so, then this code causes the p2p thread to buffer the results in a `VecDeque` outside the channel and continuously try to forward it to the relayer until the backpressure subsides.  Once backpressure subsides, the p2p thread can resume downloading blocks as quickly as possible.

There's some related code in the downloader that ensures that a block doesn't get re-downloaded more than XXX seconds.  This is in place because even with backpressure control, the p2p network doesn't know if the relayer is sitting on blocks that are yet to be processed.  Depending on your feedback, I could remove this code and simply have the p2p thread keep its own tabs on which blocks are in-flight to the relayer, and have it tell the downloader to not download them again even though they haven't reached the disk yet.

Reply:
> Depending on your feedback, I could remove this code and simply have the p2p thread keep its own tabs on which blocks are in-flight to the relayer, and have it tell the downloader to not download them again even though they haven't reached the disk yet.

Actually, I think I will just do this now.  My neon follower re-downloads over 1,000 blocks during spin-up, which is not okay.